CMakeLists.txt

```diff
@@ -257,128 +268,56 @@ endif()
 #   ENSMALLEN_INCLUDE_DIR - include directory for ensmallen
 #   STB_IMAGE_INCLUDE_DIR - include directory for STB image library
 #   MATHJAX_ROOT - root of MathJax installation
-find_package(Armadillo "${ARMADILLO_VERSION}" REQUIRED)
 
+# Remove support for older 32bit system, if we need to have 32bit system
+# Since we are cross compiling, We will download it into the deps folder,
+find_package(Armadillo "${ARMADILLO_VERSION}")
+if(NOT ARMADILLO_FOUND AND (NOT DISABLE_DOWNLOADS))
+  get_deps(http://files.mlpack.org/armadillo-10.3.0.tar.gz armadillo armadillo-10.3.0.tar.gz)
+  set(ARMADILLO_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+  find_package(Armadillo REQUIRED)
+endif()
 # Include directories for the previous dependencies.
-set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS})
-set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIR})
+
+# We do not have to link with Armadillo if we are cross compiling
+if(CMAKE_CROSSCOMPILING)
+  get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v0.3.13/OpenBLAS-0.3.13.tar.gz OpenBLAS OpenBLAS-0.3.13.tar.gz)
+  find_library(GFORTRAN NAMES libgfortran.a PATHS ${CMAKE_SYSROOT}/usr/lib)
+  find_library(PTHREAD NAMES libpthread.a PATHS ${CMAKE_SYSROOT}/usr/lib)
+  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${OPENBLAS_LIBRARIES} ${GFORTRAN} ${PTHREAD})
+else()
+  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARY})
```

Comment:
```suggestion
  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
```
This is a bit of an important point: Armadillo can be configured with or without the "wrapper"; this is controlled by the `ARMA_USE_WRAPPER` macro in `include/armadillo_bits/config.hpp`.  If `ARMA_USE_WRAPPER` is set, then we can link against only `libarmadillo.so` (or whatever the object is called).  If `ARMA_USE_WRAPPER` is not set---or if `ARMA_DONT_USE_WRAPPER` is set, then we have to link against all the dependencies of Armadillo instead.  Depending on the user's configuration, this could be more than just OpenBLAS; please take a look at the details of `FindArmadillo.cmake` to understand what's going on. :+1:

In the case where the autodownloader is being used, `ARMA_USE_WRAPPER` will not be set, as you are just using the sources directly in `include/` from the unpacked tarball.  But what happens if the autodownloader is not being used, and `ARMA_USE_WRAPPER` is true?  In this case, I suppose we need to find the statically-linked Armadillo library (`libarmadillo.a` or similar).

I think maybe we want to put some of the logic here into `FindArmadillo.cmake`.  Really the only difference between `CMAKE_CROSSCOMPILING` being true or false is whether we want static libraries or not to link against.  So perhaps we should modify `FindArmadillo.cmake` to take the variable `BUILD_SHARED_LIBS` into account, and if it's `OFF`, then search for static libraries (and also add gfortran and pthread).

Ideally, we should have as little as possible specific to `CMAKE_CROSSCOMPILING`---really that just controls whether or not we want static linking and what the compiler/toolchain is.  But our actual CMake logic should be separated from that.